text/0000-lwc-signals.md

```diff
@@ -0,0 +1,373 @@
+    count = signal(0);
+}
+```
+
+```html
+<!-- ParentComponent's template -->
+<template>
+    <!-- Pass the value of the signal, not the signal itself -->
+    <child-component count={count.value}></child-component>
+</template>
+```
+
+The child component simply receives a number as a property and uses it as such, unaware of the signal mechanism in the parent:
+
+```javascript
+export default class ChildComponent extends LightningElement {
+    @api count;
+}
+```
+
+This ensures that child components remain agnostic of the parent's state management implementation, promoting better component encapsulation and reuse.
+
+In terms of optimizations, the compiler can detect when a signal is passed down to a child component because of the `.value` in the template interpolation, and generate code to subscribe to the signal's changes. This would allow the parent component to update the child's public property `count` without a full re-render, leading to better performance. On the receiving end, because the child component's `count` property is not a signal, but a value, it will react to changes as usual, without any special handling.
+
+### Object Reactivity
+
+Objects can be wrapped in Signals to make them reactive. However, changes to nested properties must be handled by replacing the entire object:
+
+```javascript
+import { signal } from 'signals';
+
+export default class ExampleComponent extends LightningElement {
+    position = signal({ x: 0, y: 0 });
+
+    move(x, y) {
+        this.position.value = { ...this.position.value, x, y };
+    }
+}
+```
+
+### Connecting to External Data Stores
+
+Components can integrate with external stores by using a pattern similar to the current `@wire`` mechanism, but with greater flexibility:
+
+```javascript
+import { RecordStore } from 'c/myStore';
+
+export default class ExampleComponent extends LightningElement {
+    record = new RecordStore("001XXXXXXXXXXXXXXX");
+}
+```
+
+```html
+<template>
+    <p>{record.value.Name}</p>
+</template>
+```
+
+__Note:__ _since the record might not be ready, it might require some additional logic to handle the case when the record is not available. But this is the same they do today with `@wire`._
+
+### Reactive API Properties
+
+The example above shows how to integrate with external stores using a static value. But what if the value is actually a property of the component that might change over time? To facilitate interaction with external stores, LWC will provide access to reactive API properties:
+
+```javascript
+export default class ExampleComponent extends LightningElement {
+    @api recordId;
+    record = new RecordStore(this.$api.recordId);
+}
+```
+
+In this design, `this.$api` is a reactive wrapper around the component's API properties, enabling stores to react to changes in properties like recordId. This reensambles the internal mechanism used today in LWC via `vm.cmpProps` that is not exposed to developers.
```

Comment:
yes! the argument would be a signal-like object created by LWC.

Reply:
Would `RecordStore` be expected to do any `.unsubscribe` cleanup when the component is disconnected/disposed?

Reply:
I would update the API to remove the unsubscribe all together to match what other frameworks are doing. As for the development experience, they should try to unsubscribe, but LWC would not track any of that. When LWC uses (internally) an store, it would unsubscribe accordingly

Reply:
> reensambles

?

Reply:
> `new RecordStore(this.$api.recordId)`

How does this allow the `RecordStore` to react to changes in `recordId`? It's receiving the value, not the signal, correct? Or is `$api` an object where the props are strings and the values are signals?